queryable-spec.ts

```diff
@@ -0,0 +1,424 @@
+   * This is used to determine the I/O cost.
+   */
+  requestTime: number;
+  /**
+   * Custom properties
+   */
+  [key: string]: any;
+}
+
+/**
+ * QueryOperationOrder represents an ordering of the results of a given
+ * query operation. QueryOperationOrder objects can be returned by
+ * implementations of both the Filterable (quad orderings) and Queryable
+ * (bindings orderings) interfaces. Furthermore, QueryOperationObjects are
+ * used to represent both available orderings (i.e. orderings that may be
+ * requested by callers) and returned orderings (i.e. orderings followed
+ * by the returned iterators).
+ */
+interface QueryOperationOrder<T extends TermName | RDF.Variable> {
+  cost: QueryOperationCost;
+  terms: { term: T, direction: 'asc' | 'desc' }[];
+}
+
+
+/******************************************************************************
+                              FILTERABLE SOURCE
+ *****************************************************************************/
+
+
+/**
+ * QueryResultMetadataOptions is an abstract interface that represents a generic 
+ * expression over a stream of quads. 
+ */
+interface Expression {
+  /**
+   * Value that identifies the concrete interface of the expression, since the
+   * Expression itself is not directly instantiated. Possible values include 
+   * "operator" and "term".
+   */
+  expressionType: string;
+};
+
+/**
+ * An OperatorExpression is represents an expression that applies a given
+ * operator on given sub-expressions.
+ * 
+ * The WebIDL definition of the Filterable spec contains a list of supported 
+ * operators: https://rdf.js.org/query-spec/#expression-operators 
+ */
+interface OperatorExpression extends Expression {
+  
+  /** 
+   * Contains the constant "operator". 
+   */
+  expressionType: 'operator';
+
+  /**
+   * Value that identifies an operator. Possible values can be found in the 
+   * list of operators.
+   */
+  operator: string;
+
+  /**
+   * Array of Expression's on to which the given operator applies. The length
+   * of this array depends on the operator.
+   */
+  args: Expression[];
+};
+
+/**
+ * A TermExpression is an expression that contains a Term.
+ */
+interface TermExpression {
+  
+  /**
+   * The constant "term".
+   */
+  expressionType: 'term';
+
+  /**
+   * a Term.
+   */
+  term: RDF.Term;
+}
+
+/**
+ * ExpressionFactory enables expressions to be created in an idiomatic manner.
+ */
+interface ExpressionFactory {
+  
+  /**
+   * Creates a new OperatorExpression instance for the given operator and array of arguments.
+   */
+  operatorExpression(operator: string, args: Expression[]): OperatorExpression;
+
+  /**
+   * Creates a new TermExpression instance for the given term.
+   */
+  termExpression(term: RDF.Term): TermExpression;
+};
+
+
+/**
+ * QueryResultMetadataCount is part of the QueryResultMetadata interface to
+ * represent metadata about the number of quads in the result stream.
+ */
+interface FilterableResultMetadataCount {
```

Comment:
```suggestion
interface FilterableResultMetadataCardinality {
```
"cardinality" is mostly used to refer to this.